Window class

Zydak opened the pull request:
Implementation of small wrapper around GLFW

Zydak committed:
Window class

Zydak committed:
Typo

Zydak committed:
Merge branch 'devel' into Zydak-Window

Tnkln99 commented:
Can you make a separate module for the window? In general core will be included in modules that wont use window at all, only executables will use window, it can be nice too have its own module.

Tnkln99 commented:
I really like the fact that you added tests :)

Zydak committed:
[Window/Core][Window Wrapper] Moved window from core to it's separate module

Zydak commented:
Done

chursinm commented on Engine/Modules/Window/Source/Window.cpp:
```diff
@@ -0,0 +1,80 @@
+	}
+
+	void Window::resizeCallback(GLFWwindow* window, int width, int height) {
+		Window* userPointer   = (Window*)(glfwGetWindowUserPointer(window));
```
I'm not sure if you can cast `GLFWwindow*` to your `Window*` custom class. Can you write a test case for resize?

Zydak replied:
sure thing, didn't realize that's our code style, and I personally prefer C style casting since it's shorter

chursinm replied:
sorry, just edit my comment.

Zydak replied:
why wouldn't I be able to do that? It's just a void* that gets set to whatever I want to, then I just recast it to whatever it was set.

Zydak replied:
But now that you mention it, I realized I forgot to even bind the callback to a window. Fixed that in the newest commit. Resizing works fine. You can check that by running `SynergonEngineSamples` and just resizing the window

chursinm replied:
Ups, just misread the code. Resolved.

MrTitanHearted reviewed (changes requested):
I think you should change some parts:
Polling events are not for just one window but for all windows. Thus, you should make it a static method, not a member method: static void PollEvents();. Thereby, window.pollEvents(); -> Synergon::Window::PollEvents();
You don't have to use glfwSetWindowUserPointer every frame! Just do it once when initializing the window in the constructor

And this one is not necessarily required right now, but we need an even queue/pipeline/messenger/listener/observer/whatever you want to call thingy to manage events. Therefore, please look for things like that!

Zydak committed:
[Window][Window Wrapper] no C style cast + fixed the callback

Zydak committed:
[Window][Window Wrapper] pollEvents is now static

Zydak commented:
Fixed all the stuff you mentioned. And for the event system I thought about something like this:

```
void SetKeyCallback(int keyCode, uint32_t layer, std::function<bool(...Some Args...)> function)
```

So if you want to you can just set this callback to whatever function you want and it will be called when the specified `keyCode` is pressed. The callback also returns bool to signal whether the event has been consumed or not, if it was, the key press won't be propagated down to the other layers. Does that sound good?

Tnkln99 commented:
Window.hpp should not include glfw headers, this hould be included from source file so that other module that uses Window module does not need to include glfw, glfw specific classes that is on header must be forward declared
and so "target_link_libraries(SynergonWindow PUBLIC glfw)" should be private

Zydak commented:
Sure, then I'll add `target_link_libraries(glfw)` to `RHI` since it will need to link to glfw to create VkSurface. Is that fine?

MrTitanHearted commented:
> Sure, then I'll add `target_link_libraries(glfw)` to `RHI` since it will need to link to glfw to create VkSurface. Is that fine?

Sorry, but no. I want you to create something like this: `typedef void* NativeWindowHandle; and NativeWindowHandle window.getNativeWindowHandle();` which will return HWND in case of Windows, and maybe x11/wayland window handle in Linux and, the alternative one in Macos, respectively. The necessary vulkan extensions can be chosen in the rhi implementation and thus, there is no need to link glfw to rhi

>  And for the event system, I thought about something like this:

That approach works fine for now, but I still think we need a proper event system. Something like this:
* every event should be implemented Synergon::IEvent which is to describe the event
* every event is parsed in the native code itself (meaning window events are parsed in the Window Module)
* We will give callbacks like this `eventManager.addCallback(Callback&& c);` where c  takes single class thatis based off IEvent, which we can enforce using template programming
* Then we can do something like this for the processing `eventManager.process(WindowResize{400, 600});`

what I mean by all of this is that, I think we need more centralized event system where every module will parse their events and put it into a single queue that then we can use them to process on using callbacks. It does not have to be exactly like I described, but I think it needs to be more proper system than just adding a callback just for window stuff. But once again, I think this works fine as of now

Zydak commented:
> The necessary vulkan extensions can be chosen in the rhi implementation and thus, there is no need to link glfw to rhi

I'm not really sure what you mean by vulkan extensions? I meant that to create `VkSurface`, which is needed for swapchain in vulkan, we need to call `glfwCreateWindowSurface(...)`. So we either link to vulkan in `Window` module and create it in `Window`, or we link to GLFW in `RHI` module and create it in `RHI`. The point is we need both GLFW and Vulkan in one of these modules.

Tnkln99 commented:
Actually we don't need the window module on rhi, they should be separate, rhi is not a rendering module, it is a way to communicate with the gpu only (a app that will not use windowing and just want to use compute shaders to do matrices multiplications could use rhi module). glfwCreateWindowSurface is one way to create the surface, there are other ways.
Window -> app that has window <- Rpi <-Rhi
Rhi -> a module that wants to use gpu but wont have rendering

Tnkln99 commented:
> * every event should be implemented Synergon::IEvent which is to describe the event

I opened an issue if somebody wants to implement it
https://github.com/SynergonTeam/SynergonEngine/issues/15

Zydak committed:
[Window][GLFW] Window.hpp no longer has glfw headers

Zydak commented:
> glfwCreateWindowSurface is one way to create the surface, there are other ways.

Oh ok, didn't realize that.

I changed it so that now GLFW is linked privately to Window module and Window.hpp no longer has any GLFW headers. Also added the `typedef void* NativeWindowHandle;` and using different functions to get the handle on different platforms, but I can't test it since I have neither mac or linux.

Zydak committed:
[Window][Tests] Fixed tests
